File: bindings/android/hello-scion/src/main/kotlin/com/anapaya/scion/http3/hello/HelloScion.kt
// Copyright 2026 Anapaya Systems

package com.anapaya.scion.http3.hello

// ANCHOR: full-sample
import android.content.Context
import com.anapaya.scion.http3.ScionAddress
import com.anapaya.scion.http3.ScionHttp3Client
import com.anapaya.scion.http3.ScionHttp3Request
import com.anapaya.scion.http3.TrustAnchors
import java.io.Closeable
import kotlin.time.Duration.Companion.seconds

/**
 * One HTTP/3 request over SCION.
 */
class HelloScion(
    context: Context,
    private val network: LocalNetwork,
) : Closeable {
    /** What the server answered. A non-2xx status arrives here, not as an exception. */
    data class Reply(
        val code: Int,
        val body: String,
    )

    // ANCHOR: build-client
    private val client =
        ScionHttp3Client
            .Builder(context)
            // Where the client discovers its SCION connectivity.
            .endhostApi(network.endhostApiUrl)
            .authToken(network.authToken)
            .trust(TrustAnchors.pinned(network.caPem.toByteArray()))
            .connectTimeout(30.seconds)
            .requestTimeout(60.seconds)
            .build()
    // ANCHOR_END: build-client

    // ANCHOR: request

    /** `GET /hello`, which the test server answers with `world`. */
    suspend fun hello(): Reply {
        val request =
            ScionHttp3Request
                .Builder()
                .url("${network.baseUrl}/hello")
                // The test network publishes no records for its server, so address it directly.
                .target(ScionAddress.parse(network.target))
                .get()
                .build()

        return client.newCall(request).execute().use { response ->
            Reply(response.code, response.body.string())
        }
    }
    // ANCHOR_END: request

    // ANCHOR: lifetime

    /**
     * Releases the client's connections.
     *
     * This returns at once and leaves the shutdown to finish on its own. Await it with
     * [ScionHttp3Client.shutdown] instead where that matters.
     */
    override fun close() {
        client.close()
    }
    // ANCHOR_END: lifetime
}
// ANCHOR_END: full-sample

File: bindings/apple/hello-scion/HelloScion/HelloScion.swift
// Copyright 2026 Anapaya Systems

// ANCHOR: full-sample
import Foundation
import ScionHTTP3

/// One HTTP/3 request over SCION.
final class HelloScion: Sendable {
    struct Reply {
        let code: Int
        let body: String
    }

    private let network: LocalNetwork
    private let client: ScionHttp3Client

    // ANCHOR: build-client
    init(network: LocalNetwork) throws {
        self.network = network
        var configuration = ScionHttp3Client.Configuration(
            // Where the client discovers its SCION connectivity.
            endhostApi: network.endhostApiUrl,
            authToken: network.authToken)
        configuration.trust = try .pinned(Data(network.caPem.utf8))
        configuration.connectTimeout = 30
        configuration.requestTimeout = 60
        client = try ScionHttp3Client(configuration: configuration)
    }
    // ANCHOR_END: build-client

    // ANCHOR: request
    /// `GET /hello`, which the test server answers with `world`.
    func hello() async throws -> Reply {
        var request = ScionHttp3Request(url: network.baseUrl + "/hello")
        // The test network publishes no records for its server, so address it directly.
        request.target = try ScionAddress(network.target)

        let response = try await client.execute(request)
        return Reply(code: response.code, body: try await response.body.string())
    }
    // ANCHOR_END: request

    // ANCHOR: lifetime
    /// Releases the client's connections.
    ///
    /// Waits until each connection has told its peer it is going away. A client that is dropped
    /// without this still closes its connections in the background and whenever ARC gets to it.
    func shutdown() async {
        await client.shutdown()
    }
    // ANCHOR_END: lifetime
}
// ANCHOR_END: full-sample
